### PR TITLE
Expand SetCipherList()

### PR DESCRIPTION
**Pull request by julek-wolfssl**
- support disabling ciphersuites starting from the default list
- BIO_eof returns 1 when no more data remains in the BIO and 0 when there is data pending

Tested with `./configure --enable-openvpn CPPFLAGS=-DWOLFSSL_STATIC_RSA -DWOLFSSL_STATIC_PSK -DWOLFSSL_STATIC_DH --enable-nullcipher --enable-arc4 --enable-anon --enable-dh --enable-psk --enable-camellia --enable-dtls --enable-aesccm` to make sure all ciphersuites get added correctly.

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -23198,9 +23205,89 @@ int SetCipherList(WOLFSSL_CTX* ctx, Suites* suites, const char* list)
...
+                allowing = 0;
+                current++;
+                length--;
+            }
+        }
+
         XSTRNCPY(name, current, length);
         name[(length == sizeof(name)) ? length - 1 : length] = 0;
 
+        if (XSTRCMP(name, "DEFAULT") == 0) {
```
Comment (dgarske) [inline src/internal.c:23219]: Should this new code be wrapped with `OPENSSL_EXTRA` as to not grow code size?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/internal.c:23219]: Good idea. Especially with all the strings. I'll try to refactor the suite selection a bit as well.

**Review by dgarske — CHANGES_REQUESTED** [review]
Looks great. Nice refactor. Note: I am not a fan of using `TRUE` to avoid confusing bool types, but I don't see an issue with it because of how we set it up by default in `types.h`.

Getting a new error with the multi-test:

```
./configure '--enable-all' '--enable-sp-math-all' 'CC=gcc' 'CPPFLAGS=-std=c99 -pedantic -Wvla -Wdeclaration-after-statement -DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK'
make
export WOLFSSL_OPENSSL_TEST=1
./scripts/openssl.test 
...
trying wolfSSL cipher suite DHE-PSK-AES256-GCM-SHA384
Matched to OpenSSL suite support
#
# ./examples/client/client -p 61148 -g -r -l DHE-PSK-AES256-GCM-SHA384 -s  "-c/tmp/wolfssl_test_workdir.1907731/wolfssl/./scripts/../certs/client-cert.pem" "-k/tmp/wolfssl_test_workdir.1907731/wolfssl/./scripts/../certs/client-key.pem" "-A/tmp/wolfssl_test_workdir.1907731/wolfssl/./scripts/../certs/ca-cert.pem" 
wolfSSL_connect error -313, received alert fatal error
wolfSSL error: wolfSSL_connect failed
client failed! Suite = DHE-PSK-AES256-GCM-SHA384 version = 
```

**Comment by julek-wolfssl** [comment]
The error with `./scripts/openssl.test` was due to `anonymous_sa_algo` being added to the list of signature algorithms even if it wasn't present in the ciphersuites. This is now changed so that it only appears when an anonymous ciphersuite is present.

**Review by dgarske — APPROVED** [review]

**Pull request merged**